File: server/src/main/java/dev/ultreon/quantum/network/system/TcpConnection.java
package dev.ultreon.quantum.network.system;

import com.badlogic.gdx.utils.Pool;
import com.esotericsoftware.kryo.KryoException;
import com.esotericsoftware.kryonet.Connection;
import com.esotericsoftware.kryonet.FrameworkMessage;
import com.esotericsoftware.kryonet.KryoNetException;
import com.esotericsoftware.kryonet.Listener;
import dev.ultreon.quantum.network.PacketContext;
import dev.ultreon.quantum.network.PacketData;
import dev.ultreon.quantum.network.PacketHandler;
import dev.ultreon.quantum.network.PacketListener;
import dev.ultreon.quantum.network.packets.Packet;
import dev.ultreon.quantum.network.stage.PacketStage;
import dev.ultreon.quantum.network.stage.PacketStages;
import dev.ultreon.quantum.server.CloseCodes;
import dev.ultreon.quantum.server.QuantumServer;
import dev.ultreon.quantum.server.player.ServerPlayer;
import dev.ultreon.quantum.util.Env;
import dev.ultreon.quantum.util.Result;
import org.apache.commons.collections4.queue.SynchronizedQueue;
import org.jetbrains.annotations.Nullable;

import java.io.Closeable;
import java.io.IOException;
import java.nio.channels.ClosedChannelException;
import java.util.ArrayDeque;
import java.util.Queue;
import java.util.concurrent.Executor;

public abstract class TcpConnection<OurHandler extends PacketHandler, TheirHandler extends PacketHandler> extends Listener implements IConnection<OurHandler, TheirHandler>, Closeable {
    public static final Pool<Long> sequencePool = new Pool<>() {
        private long next = 0;

        @Override
        protected Long newObject() {
            return next++;
        }
    };
    private final Queue<Packet<? extends TheirHandler>> packetQueue = SynchronizedQueue.synchronizedQueue(new ArrayDeque<>());
    private final Thread senderThread;
    private final com.esotericsoftware.kryonet.Connection connection;
    private final Executor executor;
    private boolean compressed;
    private PacketData<OurHandler> ourPacketData;
    private PacketData<TheirHandler> theirPacketData;
    private OurHandler handler;
    private boolean readOnly;
    protected long ping = 0;
    private boolean loggingIn = true;

    public TcpConnection(com.esotericsoftware.kryonet.Connection connection, Executor executor) {
        this.connection = connection;
        this.executor = executor;

        connection.addListener(this);

        senderThread = new Thread(() -> {
            try {
                this.sender();
            } catch (KryoException e) {
                LOGGER.error("Failed to send packet", e);
                try {
                    this.close();
                } catch (IOException ex) {
                    LOGGER.error("Failed to close connection", ex);
                }
            } catch (ClosedChannelException e) {
                // Ignored
            } catch (IOException e) {
                this.disconnect(CloseCodes.NO_STATUS_CODE.getCode(), "Connection interrupted!");
            }
        });

        senderThread.setDaemon(true);
    }

    public static void handleReply(long sequenceId) {
        sequencePool.free(sequenceId);
    }

    private void sender() throws IOException {
        while (connection.isConnected() && isRunning()) {
            Packet<?> packet = packetQueue.poll();
            if (isReadOnly()) return;
            if (packet != null) {
                this.connection.sendTCP(packet);
            }
        }
    }

    protected abstract boolean isRunning();

    protected abstract ServerPlayer getPlayer();

    private void iLuvGenerics(Packet<OurHandler> decode, ServerPlayer player) {
        ourPacketData.handle(decode, new PacketContext(player, this, Env.SERVER), this.handler);
    }

    @Override
    public void send(Packet<? extends TheirHandler> packet) {
        if (this.readOnly)
            throw new ReadOnlyConnectionException();
        if (theirPacketData.getId(packet) < 0)
            throw new IllegalArgumentException("Invalid packet: " + packet.getClass().getName());

        if (handler.isAsync()) {
            this.packetQueue.add(packet);
        } else {
            this.connection.sendTCP(packet);
        }
    }

    @Override
    @Deprecated
    public void send(Packet<? extends TheirHandler> packet, @Nullable PacketListener resultListener) {
        if (this.readOnly)
            throw new ReadOnlyConnectionException();
        if (theirPacketData.getId(packet) < 0)
            throw new IllegalArgumentException("Invalid packet: " + packet.getClass().getName());


        this.packetQueue.add(packet);
    }

    public void setCompressed(boolean compressed) {
        this.compressed = compressed;
    }

    @Override
    public boolean isCompressed() {
        return compressed;
    }

    @Override
    public void disconnect(int code, String message) {
        QuantumServer.LOGGER.info("Disconnecting {}: {}", this.getPlayer().getName(), message);
        this.connection.sendTCP(this.getDisconnectPacket(code, message));

        try {
            connection.close();
        } catch (KryoNetException e) {
            Result<Void> voidResult = this.on3rdPartyDisconnect(CloseCodes.NO_STATUS_CODE.getCode(), e.getMessage());
            if (voidResult.isFailure())
                e.addSuppressed(voidResult.getFailure());

            QuantumServer.LOGGER.error("Failed to close socket", e);
        }
    }

    @Override
    public Result<Void> on3rdPartyDisconnect(int statusCode, String message) {
        return Result.ok();
    }

    @Override
    @SuppressWarnings("unchecked")
    public void received(com.esotericsoftware.kryonet.Connection connection, Object object) {
        super.received(connection, object);

        try {
            if (object instanceof Packet) {
                System.out.println("Received packet: " + object.getClass().getName());
                iLuvGenerics((Packet<OurHandler>) object, getPlayer());
            } else if (object instanceof FrameworkMessage.KeepAlive) {
                connection.sendTCP(new FrameworkMessage.KeepAlive());
            } else {
                connection.close();
                QuantumServer server = this.getPlayer().getServer();
                if (server != null) {
                    server.onDisconnected(this.getPlayer(), "Connection closed!");
                } else {
                    LOGGER.error("Received object of type {} from {} but no server is running!", object.getClass().getName(), connection.getRemoteAddressTCP());
                }
            }
        } catch (Exception e) {
            LOGGER.error("Failed to handle packet", e);
        }
    }

    @Override
    public void disconnected(Connection connection) {
        this.on3rdPartyDisconnect(CloseCodes.NO_STATUS_CODE.getCode(), "Connection closed!");
    }

    protected abstract Packet<TheirHandler> getDisconnectPacket(int code, String message);

    @Override
    public void queue(Runnable handler) {
        this.executor.execute(handler);
    }

    @Override
    public void start() {
        senderThread.start();
    }

    @Override
    public void moveTo(PacketStage stage, OurHandler handler) {
        this.ourPacketData = this.getOurData(stage);
        this.theirPacketData = this.getTheirData(stage);

        this.loggingIn = stage == PacketStages.LOGIN;

        this.handler = handler;
    }

    @Override
    public boolean isConnected() {
        return connection.isConnected();
    }

    @Override
    public boolean isMemoryConnection() {
        return false;
    }

    protected abstract PacketData<OurHandler> getOurData(PacketStage stage);

    protected abstract PacketData<TheirHandler> getTheirData(PacketStage stage);

    @Override
    public void close() throws IOException {
        connection.close();
    }

    @Override
    public void setReadOnly() {
        this.readOnly = true;
    }

    @Override
    public void setPlayer(ServerPlayer player) {

    }

    @Override
    public long getPing() {
        return ping;
    }

    @Override
    public void onPing(long ping) {

    }

    @Override
    public boolean isLoggingIn() {
        return loggingIn;
    }

    public boolean isReadOnly() {
        return readOnly;
    }

    protected com.esotericsoftware.kryonet.Connection getConnection() {
        return connection;
    }
}

File: server/src/main/java/dev/ultreon/quantum/network/system/TcpConnection.kt
@file:Suppress("UNCHECKED_CAST")

package dev.ultreon.quantum.network.system

import com.badlogic.gdx.utils.Pool
import com.esotericsoftware.kryonet.Connection
import com.esotericsoftware.kryonet.FrameworkMessage
import com.esotericsoftware.kryonet.KryoNetException
import com.esotericsoftware.kryonet.Listener
import com.google.common.collect.Queues
import com.google.errorprone.annotations.CanIgnoreReturnValue
import com.sun.jdi.connect.spi.ClosedConnectionException
import dev.ultreon.quantum.network.PacketContext
import dev.ultreon.quantum.network.PacketData
import dev.ultreon.quantum.network.PacketHandler
import dev.ultreon.quantum.network.PacketListener
import dev.ultreon.quantum.network.packets.Packet
import dev.ultreon.quantum.network.stage.PacketStage
import dev.ultreon.quantum.server.QuantumServer
import dev.ultreon.quantum.server.player.ServerPlayer
import dev.ultreon.quantum.util.Env
import dev.ultreon.quantum.util.Result
import java.io.Closeable
import java.io.IOException
import java.nio.channels.ClosedChannelException
import java.util.*
import java.util.concurrent.Executor

abstract class TcpConnection<OurHandler : PacketHandler, TheirHandler : PacketHandler>(
  protected val connection: Connection, private val executor: Executor
) :
  Listener(), IConnection<OurHandler, TheirHandler>,
  Closeable {
  private val packetQueue: Queue<Packet<out TheirHandler>> = Queues.synchronizedQueue(ArrayDeque())
  private val senderThread: Thread
  override var isCompressed: Boolean = false
  private var ourPacketData: PacketData<OurHandler>? = null
  private var theirPacketData: PacketData<TheirHandler>? = null
  private var handler: OurHandler? = null
  var isReadOnly: Boolean = false
    private set
  override var ping: Long = 0
    protected set

  init {
    connection.addListener(this)

    senderThread = Thread {
      try {
        this.sender()
      } catch (e: ClosedChannelException) {
        // Ignored
      } catch (_: ClosedConnectionException) {
      } catch (e: IOException) {
        this.disconnect("Connection interrupted!")
      }
    }

    senderThread.isDaemon = true
  }

  @Throws(IOException::class)
  private fun sender() {
    while (connection.isConnected && isRunning) {
      val packet: Packet<*>? = packetQueue.poll()
      if (isReadOnly) return
      if (packet != null) {
        connection.sendTCP(packet)
      }
    }
  }

  protected abstract val isRunning: Boolean

  protected abstract fun getPlayer(): ServerPlayer

  private fun iLuvGenerics(decode: Packet<OurHandler>, player: ServerPlayer) {
    ourPacketData!!.handle(
      decode, PacketContext(player, this, Env.SERVER),
      this.handler
    )
  }

  @CanIgnoreReturnValue
  override fun send(packet: Packet<out TheirHandler?>) {
    if (this.isReadOnly) throw ReadOnlyConnectionException()
    require(theirPacketData!!.getId(packet) >= 0) { "Invalid packet: " + packet.javaClass.name }

    packetQueue.add(packet)
  }

  @Deprecated("")
  override fun send(packet: Packet<out TheirHandler>, resultListener: PacketListener?) {
    if (this.isReadOnly) throw ReadOnlyConnectionException()
    require(theirPacketData!!.getId(packet) >= 0) { "Invalid packet: " + packet.javaClass.name }


    packetQueue.add(packet)
  }

  override fun disconnect(message: String) {
    this.send(this.getDisconnectPacket(message))

    try {
      connection.close()
    } catch (e: KryoNetException) {
      val voidResult = this.on3rdPartyDisconnect(e.message ?: "null")
      if (voidResult.isFailure) e.addSuppressed(voidResult.failure)

      QuantumServer.LOGGER.error("Failed to close socket", e)
    }
  }

  override fun on3rdPartyDisconnect(message: String): Result<Void> {
    return Result.ok()
  }

  override fun received(connection: Connection, `object`: Any) {
    super.received(connection, `object`)

    try {
      when (`object`) {
        is Packet<*> -> {
          iLuvGenerics(`object` as Packet<OurHandler>, getPlayer())
        }

        is FrameworkMessage.KeepAlive -> {
          connection.sendTCP(FrameworkMessage.KeepAlive())
        }

        else -> {
          connection.close()
        }
      }
    } catch (e: Exception) {
      IConnection.logger.error("Failed to handle packet", e)
    }
  }

  override fun disconnected(connection: Connection) {
    this.on3rdPartyDisconnect("Connection closed!")
  }

  protected abstract fun getDisconnectPacket(message: String): Packet<TheirHandler>

  override fun queue(handler: Runnable) {
    executor.execute(handler)
  }

  override fun start() {
    senderThread.start()
  }

  override fun moveTo(stage: PacketStage, handler: OurHandler) {
    this.ourPacketData = this.getOurData(stage)
    this.theirPacketData = this.getTheirData(stage)

    this.handler = handler
  }

  override val isConnected: Boolean
    get() = connection.isConnected

  override val isMemoryConnection: Boolean
    get() = false

  protected abstract fun getOurData(stage: PacketStage): PacketData<OurHandler>

  protected abstract fun getTheirData(stage: PacketStage): PacketData<TheirHandler>

  @Throws(IOException::class)
  override fun close() {
    connection.close()
  }

  override fun setReadOnly() {
    this.isReadOnly = true
  }

  override fun setPlayer(player: ServerPlayer?) {
  }

  override fun onPing(ping: Long) {
  }

  companion object {
    val sequencePool: Pool<Long> = object : Pool<Long>() {
      private var next: Long = 0

      override fun newObject(): Long {
        return next++
      }
    }

    @JvmStatic
    fun handleReply(sequenceId: Long) {
      sequencePool.free(sequenceId)
    }
  }
}
